mesonbuild/interpreter/interpreterobjects.py

```diff
@@ -302,51 +307,28 @@ def is_used(self) -> bool:
-                             'become a hard error in the future.',
-                             location=self.current_node)
-            args = args[0]
-
-        if len(args) != 2:
-            raise InterpreterException("Configuration set requires 2 arguments.")
+    def __check_used(self) -> None:
         if self.used:
             raise InterpreterException("Can not set values on configuration object that has been used.")
-        name, val = args
-        if not isinstance(val, (int, str)):
-            msg = f'Setting a configuration data value to {val!r} is invalid, ' \
-                  'and will fail at configure_file(). If you are using it ' \
-                  'just to store some values, please use a dict instead.'
-            mlog.deprecation(msg, location=self.current_node)
-        desc = kwargs.get('description', None)
-        if not isinstance(name, str):
-            raise InterpreterException("First argument to set must be a string.")
-        if desc is not None and not isinstance(desc, str):
-            raise InterpreterException('Description must be a string.')
-
-        # TODO: Remove the cast once we get rid of the deprecation
-        return name, T.cast(T.Union[str, bool, int], val), desc
 
-    @noArgsFlattening
-    def set_method(self, args: T.List[TYPE_var], kwargs: TYPE_kwargs) -> None:
-        (name, val, desc) = self.validate_args(args, kwargs)
-        self.conf_data.values[name] = (val, desc)
-
-    def set_quoted_method(self, args: T.List[TYPE_var], kwargs: TYPE_kwargs) -> None:
-        (name, val, desc) = self.validate_args(args, kwargs)
-        if not isinstance(val, str):
-            raise InterpreterException("Second argument to set_quoted must be a string.")
-        escaped_val = '\\"'.join(val.split('"'))
-        self.conf_data.values[name] = ('"' + escaped_val + '"', desc)
-
-    def set10_method(self, args: T.List[TYPE_var], kwargs: TYPE_kwargs) -> None:
-        (name, val, desc) = self.validate_args(args, kwargs)
-        if val:
-            self.conf_data.values[name] = (1, desc)
-        else:
-            self.conf_data.values[name] = (0, desc)
+    @typed_pos_args('configuration_data.set', str, (str, int, bool))
+    @typed_kwargs('configuration_data.set', *_CONF_DATA_SET_KWS)
+    def set_method(self, args: T.Tuple[str, T.Union[str, int, bool]], kwargs: 'kwargs.ConfigurationDataSet') -> None:
+        self.__check_used()
+        self.conf_data.values[args[0]] = (args[1], kwargs['description'])
+
+    @typed_pos_args('configuration_data.set_quoted', str, str)
+    @typed_kwargs('configuration_data.set_quoted', *_CONF_DATA_SET_KWS)
+    def set_quoted_method(self, args: T.Tuple[str, str], kwargs: 'kwargs.ConfigurationDataSet') -> None:
+        self.__check_used()
+        escaped_val = '\\"'.join(args[1].split('"'))
+        self.conf_data.values[args[0]] = (f'"{escaped_val}"', kwargs['description'])
+
+    @typed_pos_args('configuration_data.set10', str, (int, bool))
```

Comment:
This only takes bool, not int.

Reply:
we document that we do: docs/yaml/objects/feature.yaml

Reply:
What do feature objects have to do with cfg_data? :p And nowhere in that file is `int` mentioned.

cfg_data does mention int, but only in the context of set / set_quoted, not set10. That is because those actually support defining e.g. `3`.

Reply:
from said place:
```meson
     if get_option('directx').require(host_machine.system() == 'windows',
           error_message: 'DirectX only available on Windows').allowed() then
       src += ['directx.c']
       config.set10('HAVE_DIRECTX', 1)
     endif
```

I'm not opossed to remove bool support, but I suspect since we've documented it, and we've allowed it, someone is using it; and unlike passing say `'1'` there isn't any reason that passing a literal number isn't valid, since meson does inherit a bit of the bool is int madness from C.

Reply:
There is absolutely a reason to not allow it. If we allow it, people can pass `set10('VAR', 5)` and that is interpreted as "true" same as "1". In fact, the reason it is interpreted as true is because we currently do no validation at all, you can do `config.set10('HAVE_DIRECTX', "the little brown fox jumped over the lazy dog")` and meson does a python

```python
        if val:
            self.conf_data.values[name] = (1, desc)
        else:
            self.conf_data.values[name] = (0, desc)
```

Since *any* non-empty string, and any non-zero integer, and anything that generally has contents, is true in python... this sets to 1.

In fact, this works too:

```meson
config.set10('HAVE_DIRECTX', ' ')
```
A string of length one, filled with space characters. Equates to true.

This also evaluates to true:

```meson
config.set10('HAVE_DIRECTX', -1)
```
Meson has negative integers. Python considers negative integers to be truthy, not falsy.


This was added quite recently in 2f2d99e1d88c9f14fa139d450cb6f55173d73c9a and it isn't even the set10 documentation...

I'm *strongly* opposed to documenting it as allowed, but would begrudgingly compromise on raising a deprecation warning declaring that it is an invalid construct and scheduling it for removal in 0.62.0

But as far as I'm concerned, the documentation is buggy and makes no sense. :/

If we are going to keep it and mark it as deprecated, I desperately hope no one is accidentally passing negative numbers and getting truthy results!!!

Reply:
I approve this PR once a deprecation is added for passing integers here.